### PR TITLE
upgrade webpack and dev server

### LINKED ISSUE #477: Update webpack and dev server

### PR DESCRIPTION
**Pull request by pinzhenx**
Fix #477
- Upgrade Webpack from 3 to 4
- Replace `http-server` with `webpack-dev-server`
- Support hot reloading
- Support HTTPS
- Enable source map

### CODE REVIEW

**Comment by pinzhenx** [comment]
Should we still keep `npm watch`? If anyone relied on this command before, please let me know.

**Comment by huningxin** [comment]
I prefer to keeping `npm watch`. Thanks!

**Comment by ibelem** [comment]
Thanks for the great https support automatically which helps our efficiency quite a lot!
https://localhost:8080/, https://127.0.0.1:8080/, https://hostname:8080/ works well, https://intranet_ip:8080/ need remove proxy to access.

Confirmed the hot reloading will not work for `examples` but for `src`...

**Comment by ibelem** [comment]
@BruceDai @zhaoming0 if no conflicts with CI, I think we can merge this PR.

File: webpack.config.js [hunk 1]
```diff
@@ -15,20 +15,31 @@ const config = {
...
 if (process.env.NODE_ENV === 'production') {
-  config.plugins = [
-    new webpack.DefinePlugin({ 'process.env.NODE_ENV': JSON.stringify('production') }),
-    new webpack.optimize.ModuleConcatenationPlugin(),
-    new webpack.optimize.UglifyJsPlugin({
-      compress: { warnings: false, unused: false },
-      output: { comments: false }
-    })
-  ]
+  config.mode = 'production';
```
Comment (ibelem) [inline webpack.config.js]: webpack 4.x update:
- "mode" added, need to set "production" or "development"
- No necessary to define process.env.NODE_ENV value
- ModuleConcatenationPlugin was deprecated and is now optimization.concatenateModules. It's set to on by default in production mode.
- uglifyjs-webpack-plugin was enabled by default.

**Comment by ibelem** [comment]
Thanks @pinzhenx, merge this PR.

**Comment by pinzhenx** [comment]
@ibelem Thanks for your detailed documentation

**Pull request merged**